languages/python/reference/exercise-concepts/markdown.md

```diff
@@ -0,0 +1,48 @@
+
+- `str.format()` :  The example solution uses `str.split()` to break the passed in markdown string into a list of lines broken up by the `\n` character.
+
+- `re.match()`:  Method of the `re` class that returns a `match` object with any matched values from a specified Regualr Expression or pre-compliled Regular Expression.  The example uses `re.match()` in multiple places to search for text patterns that need re-formatting or subsitituting. 
+
+- regular expressions:  As a seperate language with a seperate interpreter that is often embeded inside another programming language.  Like many other programming lanagues in use, python supports the Regular Expression language and implements a Regular Expression interpreter in the core `re` module.  Both the original code to be refactored for this exercise and the example solution use the core `re` module to access Regualr Expression functionality.
+
+- `return` keyword:  Most of the functions in the example solution specify a _return_ value using the `return` keyword.
+
+- `None`:  None (null) type.  Used to define a null variable or an object.  It is in of itself a singleton in any given python program.  The example uses `None` as an expecit return value in several functions.
+
+- booleans:  True and False of type `bopl`.  The example solution uses `True` and `False` as return values from functions that test membership in a list of values.
+
+- assignment:  The example solution uses assingment for variables and other values.
+
+- `re.sub()`:  Metho of the `re` class that replaces a Regular Expression match with a new value.  The example solution uses this method in various places to susititue _markdown_ syntax for _HTML_ syntax in the passed-in markdown text.
+
+- dictionaries:  Mapping type.  The example solution employes a dictionary to return values from the `parse_line()` function.
+
+- `str.split()`:  String method.  The example solution uses `str.split()` to split the incoming string into lines (_based on `\n`_), so that the text can be more easily processed for output.
+
+- for loops:  The example solution uses `for` loops to iterate over various function inputs.
+
+- iteration/iterable:  The example solution uses the `for _ in _` syntax to iterate over a list of lines.  This is possible because a list is an `iterable`.
+
+- conditionals:  The example solution uses `if` to check for pattern matching and membershihp conditions in differnt functions for processing different markdown patterns.
+
+- `match.group`:  A retun object of the `re.match()` method.  Returned when there are one or more Regular Expression groups specified in the RE pattern.
+
+- while loop:  Indefinite iteration with or without a conditional.  A loop that continues unless an explicit break or condition is met.  The example uses `while` loops to continually check if certain formatting conditions have been completed or met before proceeding to compliling the final returned result.
```

Comment:
Spelling: "compliling" -> "compiling". Also `while` _always_ has a conditional; if it's a non-callable object then the conditional is `bool(object)`, but there's always a conditional test involved in keeping the loop running. There's also _zero_ reason a while should appear in this example solution. Sigh.

Reply:
I'm going to omit this.  This `example.py` is making me crazy.